### PR TITLE
internal/router: handle not-found metadata as newly specified in charm API document

### PR DESCRIPTION
**Pull request by rogpeppe**
Also drive-by fix untested meta listing endpoint and add test for it.

### CODE REVIEW

**Comment by jujugui** [comment]
please test this

**Comment by frankban** [comment]
:+1: thank you!

File: internal/router/router.go [hunk 1]
```diff
@@ -112,7 +112,10 @@ func New(handlers *Handlers, resolveURL func(url *charm.URL) error) *Router {
 	return r
 }
 
-var ErrNotFound = fmt.Errorf("not found")
+var (
+	ErrNotFound     = fmt.Errorf("not found")
+	ErrDataNotFound = fmt.Errorf("metadata not found")
```
Comment (urosj) [inline internal/router/router.go:118]: +1

File: internal/router/router.go [hunk 2]
```diff
@@ -238,14 +245,20 @@ func (r *Router) serveBulkMeta(w http.ResponseWriter, req *http.Request) (interf
...
+				// URLs not found will be omitted from the result.
+				// http://tinyurl.com/o5ptfkk
+				continue
+			}
 			return nil, err
 		}
 		meta, err := r.serveMeta(url, req)
+		if err == ErrDataNotFound {
+			// The relevant data does not exist.
+			// http://tinyurl.com/o5ptfkk
```
Comment (urosj) [inline internal/router/router.go:260]: +1

**Comment by urosj** [comment]
:+1:

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**